src/EditorFeatures/Core/GoToImplementation/GoToImplementationCommandHandler.cs

```diff
@@ -34,37 +35,38 @@ internal partial class GoToImplementationCommandHandler : VSCommanding.ICommandH
             _streamingPresenters = streamingPresenters;
         }
 
-        private (Document, IFindUsagesService) GetDocumentAndService(ITextSnapshot snapshot)
-        {
-            var document = snapshot.GetOpenDocumentInCurrentContextWithChanges();
-            return (document, document?.GetLanguageService<IFindUsagesService>());
-        }
-
         public string DisplayName => EditorFeaturesResources.Go_To_Implementation;
 
         public VSCommanding.CommandState GetCommandState(GoToImplementationCommandArgs args)
         {
             // Because this is expensive to compute, we just always say yes as long as the language allows it.
-            var (document, findUsagesService) = GetDocumentAndService(args.SubjectBuffer.CurrentSnapshot);
+            var document = args.SubjectBuffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
+            var findUsagesService = document?.GetLanguageService<IFindUsagesService>();
             return findUsagesService != null
                 ? VSCommanding.CommandState.Available
                 : VSCommanding.CommandState.Unavailable;
         }
 
         public bool ExecuteCommand(GoToImplementationCommandArgs args, CommandExecutionContext context)
         {
-            var (document, findUsagesService) = GetDocumentAndService(args.SubjectBuffer.CurrentSnapshot);
-            if (findUsagesService != null)
+            using (context.OperationContext.AddScope(allowCancellation: true, EditorFeaturesResources.Locating_implementations))
             {
-                var caret = args.TextView.GetCaretPoint(args.SubjectBuffer);
-                if (caret.HasValue)
+                var subjectBuffer = args.SubjectBuffer;
+                var document = subjectBuffer.CurrentSnapshot.GetFullyLoadedOpenDocumentInCurrentContextWithChangesAsync(
+                    context.OperationContext).WaitAndGetResult(context.OperationContext.UserCancellationToken);
+                var findUsagesService = document?.GetLanguageService<IFindUsagesService>();
+                if (findUsagesService != null)
```

Comment:
this is bad.  you're doing the expensive thing firs,t then the cheap thing second.  it should be reversed.

Reply:
reversing it is not big issue even if we do that later. also, not sure except theoradical gain, not sure what we actually gain in real life in VS.